File: ios/VineTrack/App/ChemicalIntelligence/SavedChemicalEntrySource.swift
import Foundation

/// Evidence provenance shared by new V2 saves and legacy queued-upload repair.
nonisolated enum SavedChemicalEntrySource {
    static func reviewed(isManual: Bool, isMaster: Bool, intelligence: ChemicalIntelligence) -> String {
        if isManual { return "customer_entered" }
        if isMaster { return "master_catalogue" }
        return intelligence.hasEvidencedRegistration ? "register_lookup" : "label_lookup"
    }

    /// Only rewrite the three obsolete V2 spellings. Never change other legacy values.
    static func repaired(_ source: String?, intelligence: ChemicalIntelligence?) -> String? {
        switch source {
        case "manual_v2": return "customer_entered"
        case "master_catalogue_v2": return "master_catalogue"
        case "label_lookup_v2":
            return intelligence?.hasEvidencedRegistration == true ? "register_lookup" : "label_lookup"
        default: return source
        }
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/chemical/SavedChemicalEntrySource.kt
package com.rork.vinetrack.data.chemical

/** Evidence provenance for Saved Chemicals, independent of the search UI version. */
object SavedChemicalEntrySource {
    fun reviewed(isManual: Boolean, isMaster: Boolean, intelligence: ChemicalIntelligence): String = when {
        isManual -> "customer_entered"
        isMaster -> "master_catalogue"
        intelligence.hasEvidencedRegistration -> "register_lookup"
        else -> "label_lookup"
    }

    /** Repairs only obsolete V2 values; preserves other historical values as-is. */
    fun repaired(source: String?, intelligence: ChemicalIntelligence?): String? = when (source) {
        "manual_v2" -> "customer_entered"
        "master_catalogue_v2" -> "master_catalogue"
        "label_lookup_v2" -> if (intelligence?.hasEvidencedRegistration == true) "register_lookup" else "label_lookup"
        else -> source
    }
}
